File: app/src/main/java/org/wikipedia/activity/SingleFragmentActivity.java
package org.wikipedia.activity;

import android.os.Bundle;

import androidx.annotation.IdRes;
import androidx.annotation.LayoutRes;
import androidx.fragment.app.Fragment;

import org.wikipedia.R;

/**
 * Boilerplate for a FragmentActivity containing a single stack of
 * Fragments.
 */
public abstract class SingleFragmentActivity<T extends Fragment> extends BaseActivity {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayout());

        if (!isFragmentCreated()) {
            addFragment(createFragment());
        }
    }

    protected void addFragment(T fragment) {
        getSupportFragmentManager().beginTransaction().add(getContainerId(), fragment).commit();
    }

    protected abstract T createFragment();

    /** @return The Fragment added to the stack. */
    protected T getFragment() {
        //noinspection unchecked
        return (T) getSupportFragmentManager().findFragmentById(getContainerId());
    }

    /** @return The resource layout to inflate which must contain a {@link android.view.ViewGroup}
     * whose ID is {@link #getContainerId()}. */
    @LayoutRes
    protected int getLayout() {
        return R.layout.activity_single_fragment;
    }

    /** @return The resource identifier for the Fragment container. */
    @IdRes
    protected int getContainerId() {
        return R.id.fragment_container;
    }

    protected boolean isFragmentCreated() {
        return getFragment() != null;
    }

}

File: app/src/main/java/org/wikipedia/activity/SingleFragmentActivity.kt
package org.wikipedia.activity

import android.os.Bundle
import android.widget.FrameLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.commit
import org.wikipedia.R
import org.wikipedia.util.DeviceUtil

/**
 * Boilerplate for a FragmentActivity containing a single stack of Fragments.
 */
abstract class SingleFragmentActivity<T : Fragment> : BaseActivity() {
    lateinit var fragment: T

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!DeviceUtil.assertAppContext(this)) {
            return
        }

        inflateAndSetContentView()

        val currentFragment: T? = supportFragmentManager.findFragmentById(R.id.fragment_container) as T?
        if (currentFragment != null) {
            fragment = currentFragment
        } else {
            fragment = createFragment()
            supportFragmentManager.commit { add(R.id.fragment_container, fragment) }
        }
    }

    protected abstract fun createFragment(): T

    protected open fun inflateAndSetContentView() {
        setContentView(R.layout.activity_single_fragment)
    }

    fun disableFitsSystemWindows() {
        findViewById<FrameLayout>(R.id.fragment_container).fitsSystemWindows = false
    }
}
